--- vinagre-2.28.1/vinagre/vinagre-window.c
+++ vinagre-2.28.1/vinagre/vinagre-window.c
@@ -1011,8 +1011,9 @@ vinagre_window_conn_exists (VinagreWindo
       c = VINAGRE_CONNECTION (l->data);
 
       if (!strcmp (vinagre_connection_get_host (conn), vinagre_connection_get_host (c)) &&
+	  g_strcmp0 (vinagre_connection_get_username (conn), vinagre_connection_get_username (c)) == 0 &&
 	  vinagre_connection_get_port (conn) == vinagre_connection_get_port (c) &&
-	  vinagre_connection_get_protocol (conn) == vinagre_connection_get_protocol (c))
+	  g_strcmp0 (vinagre_connection_get_protocol (conn), vinagre_connection_get_protocol (c)) == 0)
 	{
 	  tab = g_object_get_data (G_OBJECT (c), VINAGRE_TAB_KEY);
 	  break;
